### PR TITLE
JETUT-34: Example intro security and style for tutorial

### PR DESCRIPTION
**Pull request by arjantijms**
Two examples have been added.

### CODE REVIEW

**Review by kito99 — CHANGES_REQUESTED** [review]

**Review by kito99 — APPROVED** [review]

**Comment by kito99** [comment]
@arjantijms  and @BalusC  I merged this for now so we can keep the ball rolling. I'm now applying these minor copy edits myself, closing the comment as I go. Once I'm done I'll have one PR with all of the minor edits to review, and then Arjan can respond to the few content-specific comments left in this PR separetly.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+}
+----
+
+This identity store only "contains" the single identity (user) "john", with password "secret1" and roles "user" and "caller". Defining this kind of identity store is often the simplest way to get started. Note that Jakarta Security does not define a simple identity store out of the box, because there are questions whether that would promote security best practices.
+
+Also note that the identity store is installed and used by the security system just by the virtue of being there; it picks up all enabled CDI beans that implement `IdentityStore`. Such beans can be enabled by the security system itself (following some configuration annotation), or can be programmatically added using the appropriate CDI APIs. Where the bean comes from doesn't matter for Jakarta Security, only the fact that it's there.
+
+
+==== Test the application
+
```
Comment (kito99) [inline src/main/asciidoc/security-refresh/security.adoc:184]: Is this the text you were talking about breaking into partials so we could repeat it?

File: src/main/asciidoc/security-refresh/security.adoc [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+Note that in Jakarta EE, internally these XML constraints are transformed into `Permission` instances and made available via a specific type of the Permission Store that we explained above. Knowledge about this transformation is only needed for very advanced use cases.
+
+The observant reader may wonder if XML is really the only option here, given the strong feelings that exist in parts of the community around XML. The answer is yes and no. Jakarta EE does define the `@RolesAllowed` annotation that could be used to replace the XML shown above, but only the legacy Enterprise Beans has specified a behaviour for this when put on an Enterprise Bean. Jakarta REST has done no such thing, although the JWT API in MicroProfile has defined this for REST resources. In Jakarta EE however this remains a vendor specific extension.
+
+There are also a number of annotations and APIs in Jakarta EE to set these kinds of constraints for individual Servlets, but those won't help us much either here.
+
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:140]: Something like a callout would be very nice indeed. If we have a good style to support such a thing I'd love to see that.

Indeed, there are still many disjointed parts, and the problem is not just that those parts exists, but that there remains resistance to fix it.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+            return new CredentialValidationResult("john", Set.of("user", "caller"));
+        }
+
+        return INVALID_RESULT;
+    }
+
+}
+----
+
+This identity store only "contains" the single identity (user) "john", with password "secret1" and roles "user" and "caller". Defining this kind of identity store is often the simplest way to get started. Note that Jakarta Security does not define a simple identity store out of the box, because there are questions whether that would promote security best practices.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:178]: > However, do you mean to say that it might promote bad security practices?''

It can be both I guess. The intend is to say that everything in Jakarta Security promotes best practices, and people are not really sure if a simple identity store fits in with those best practices.

> I also wonder if this shouldn't be a callout as well.

Yes, good candidate for a callout too.

File: src/main/asciidoc/security-refresh/security.adoc (hunk 1 above)
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:184]: Yes those are among them. There are a few others.

File: src/main/asciidoc/security-refresh/security.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (kito99) [reply src/main/asciidoc/security-refresh/security.adoc:140]: Ok. We do have styling. See: https://virtua-tech.github.io/jakartaee-tutorial-playbook/jakarta-ee-docs/9.1/security/security-webtier/security-webtier.html#_specifying_a_secure_connection (the "NOTE" sections).

This can be done with normal asciidoc:

```asciidoc
[NOTE]
In practice, Jakarta EE servers treat the `CONFIDENTIAL` and `INTEGRAL` transport guarantee values identically
```

File: src/main/asciidoc/security-refresh/security.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (kito99) [reply src/main/asciidoc/security-refresh/security.adoc:178]: Ok. Can you maybe reword this a bit, then?

File: src/main/asciidoc/security-refresh/security.adoc (hunk 1 above)
Comment (kito99) [reply src/main/asciidoc/security-refresh/security.adoc:184]: OK. FYI: https://docs.antora.org/antora/latest/page/partials/.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 4]
```diff
@@ -0,0 +1,410 @@
+= Jakarta Security
+
+Jakarta Security is the overarching security API in Jakarta EE.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:3]: I guess so. This was somewhat explained in the callout by mentioning many APIs in EE have (or historically had) their own stuff for security. I'll take a look to see why that isn't clear and/or how to rephrase that.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+
+Jakarta Security makes an important distinction between several distinct artefacts that play an important role in the security process:
+
+1. The Authentication Mechanism
+2. The Identity Store
+3. The Permission Store
+
+The first two of these are used in the authentication process:
+
+An _authentication mechanism_ is somewhat like a controller in the well-known MVC pattern; it’s the entity that interacts with the caller (typically a human) via some kind of view to collect credentials and with the model (business logic) to validate these credentials. An authentication mechanism knows about the environment that this caller uses to communicate with the server. An authentication mechanism for HTTP knows about URLs to redirect or forward to, or about response headers to send to the client. It also knows about the data coming back, such as cookies, request headers, and post data. Examples of authentication mechanisms are FORM authentication and BASIC authentication.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:26]: Could do so indeed. This one is quite neutral I think: https://en.wikipedia.org/wiki/Model–view–controller

File: src/main/asciidoc/security-refresh/security.adoc [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+An _identity store_ on its turn is more like the model in the MVC pattern. This entity strictly performs a business / data operation where credentials go in, and an identity comes out. The identity contains logic to validate said credentials, and embeds or contacts a database. This "database" contains user names, along with their credentials and (typically) roles. An identity store therefore knows nothing about the environment that this caller uses to communicate with the server; e.g. it doesn't know about HTTP or headers etc.
+Example of identity stores are services that contact SQL/JDBC databases, LDAP servers, files on the file-system, etc.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:29]: Well, the provided identity store in Jakarta Security is SQL/JDBC based, hence why I used that there. Obviously in general it could be SQL and NoSQL indeed.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+
+An _identity store_ on its turn is more like the model in the MVC pattern. This entity strictly performs a business / data operation where credentials go in, and an identity comes out. The identity contains logic to validate said credentials, and embeds or contacts a database. This "database" contains user names, along with their credentials and (typically) roles. An identity store therefore knows nothing about the environment that this caller uses to communicate with the server; e.g. it doesn't know about HTTP or headers etc.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:28]: Thanks, I like that one too :)

File: src/main/asciidoc/security-refresh/security.adoc [hunk 8]
```diff
@@ -0,0 +1,773 @@
...
+Example of identity stores are services that contact SQL/JDBC databases, LDAP servers, files on the file-system, etc.
+
+
+image::authentication_mvc.svg["Diagram illustrating the role of the authentication mechanism and identity store in an MVC like structure"]
+
+The third one is used for the authorization process:
+
+A _permission store_ is another kind of model that stores permissions, typically either globally, or per role (role based permissions). This entity then performs a business / data operation where a query and an identity go in, and a yes/no answer goes out. For instance, a query such as "can access /foo/bar" along with the identity for user "John" with roles "bar" and "kaz".
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:36]: I get it, yes that would be good to clarify there.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 9]
```diff
@@ -0,0 +1,773 @@
...
+1. The Database Identity Store
+2. The LDAP Identity Store
+
+== Custom authentication mechanisms and identity stores 
+
+When the provided authentication mechanisms and identity stores are not sufficient, we can easily define our own. The provided ones and our own ones use the exact same interfaces, and the system doesn't distinguish between calling our own one or one that was provided by the system.
+
+== Authentication mechanisms and identity stores from other APIs
+
+The Servlet specification defines the exact same Form and Basic authentication mechanisms. Authenticating with them will have the exact same result as authenticating with a Jakarta Security authentication mechanism. E.g. role checks will work exactly the same independent on which API was used to authenticate.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:62]: Agreed. Let's tone it down a little.

File: src/main/asciidoc/security-refresh/security.adoc [hunk 10]
```diff
@@ -0,0 +1,773 @@
...
+
+== Authentication mechanisms and identity stores from other APIs
+
+The Servlet specification defines the exact same Form and Basic authentication mechanisms. Authenticating with them will have the exact same result as authenticating with a Jakarta Security authentication mechanism. E.g. role checks will work exactly the same independent on which API was used to authenticate. 
+
+A Servlet authentication mechanism however will not necessarily consult a Jakarta Security identity store. This is server dependent. The identity store that is called is server dependent as well. Calling this server dependent identity store is possible from Jakarta Security, but as an advanced feature.
+
+Likewise, programmatic role checks can be done from various APIs, including Jakarta Security, Jakarta REST and Jakarta Servlet. These all return the same outcome, independent from whether authentication took place with a Jakarta Security Authentication Mechanism or a Servlet Authentication Mechanism.
+
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:67]: In reality there are even more ways :/ I just enumerated three of them. The fun of every API/spec wanting to be used standalone and not depend on anything else (or people designing it just not realising they are part of a larger framework called Jakarta EE)

File: src/main/asciidoc/security-refresh/security.adoc [hunk 11]
```diff
@@ -0,0 +1,773 @@
...
+            securityContext.getCallerPrincipal().getName() + " : " +
+            securityContext.isCallerInRole("user");
+    }
+
+}
+----
+
+This resource uses the injected Jakarta EE SecurityContext to obtain access to the current authenticated caller, which is represented by a `Principal` instance. For simplicity sake the code here does not check whether the caller principal is `null`. If the resource is also available to non-authenticated callers, this is something we have to do. The `getCallerPrincipal()` call would return `null` then.
+
+Also note that there is a Jakarta REST specific type that is also named `SecurityContext` and has similar methods as the ones we used here. From the point of view of Jakarta EE that is a deprecated type and replaced by the Jakarta Security version.
```
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:106]: There have been talks about deprecating it, but it's complicated. KARG argues that REST users have no problems, so it should not change. And other people argue that since REST is in core profile (and therefor in MicroProfile) and Jakarta Security is not, it can't be deprecated. It's a big mess in EE :(

Maybe we can eventually have wording like "in a web profile or full profile product this type should not be used" or something along those lines, but it's not there now.

File: src/main/asciidoc/security-refresh/security.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (arjantijms) [reply src/main/asciidoc/security-refresh/security.adoc:3]: Will add "Overarching here means that it strives to address the security needs of all other APIs in Jakarta EE."
